internal/schema/schema.go

```diff
@@ -355,66 +355,72 @@ func GetPublicSchema(ctx context.Context, db queries.DBTX) (Schema, error) {
 	//
 	// In the future, we should maybe create options where users can pass in a DB pool (WithPool(db) or WithConnection(db))
 	// and we can set concurrency to 1 if the passed in db is not a *sql.DB.
-	goRoutineRunner := concurrent.NewSynchronousGoRoutineRunner()
+	goroutineRunnerFactory := concurrent.NewSynchronousGoroutineRunner
 	if _, ok := db.(*sql.DB); ok {
-		goRoutineRunner = concurrent.NewGoroutineLimiter(50)
+		goroutineRunnerFactory = func() concurrent.GoroutineRunner {
+			return concurrent.NewGoroutineLimiter(50)
+		}
 	}
 
 	return (&schemaFetcher{
-		q:               queries.New(db),
-		goroutineRunner: goRoutineRunner,
+		q:                      queries.New(db),
+		goroutineRunnerFactory: goroutineRunnerFactory,
 	}).getPublicSchema(ctx)
 }
 
 type schemaFetcher struct {
-	q               *queries.Queries
-	goroutineRunner concurrent.GoRoutineRunner
+	q *queries.Queries
+	// goroutineRunnerFactory is a factory function that returns a GoroutineRunner. We need to be able to construct
+	// multiple GoroutineRunners to avoid deadlock created by circular dependencies of submitted go routines.
+	goroutineRunnerFactory func() concurrent.GoroutineRunner
 }
 
 func (s *schemaFetcher) getPublicSchema(ctx context.Context) (Schema, error) {
-	extensionsFuture, err := concurrent.SubmitFuture(ctx, s.goroutineRunner, func() ([]Extension, error) {
+	goroutineRunner := s.goroutineRunnerFactory()
```

Comment:
nit: Naming this field buildGoroutineRunner or something like that would make it make more sense when called.